src/Clash/Cores/Rgmii.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{- |
+Module      : Clash.Cores.Rgmii
+Description : Functions and types to connect an RGMII PHY to a packet stream interface.
+
+To keep this module generic, users will have to provide their own "primitive" functions:
+
+    1. delay functions to set to the proper amount of delay (which can be different for RX and TX);
+    2. iddr function to turn a single DDR (Double Data Rate) signal into 2 non-DDR signals;
+    3. oddr function to turn two non-DDR signals into a single DDR signal.
+
+Note that Clash models a DDR signal as being twice as fast, thus both facilitating
+and requiring type-level separation between the two "clock domains".
+-}
+module Clash.Cores.Rgmii (
+  RgmiiRxChannel (..),
+  RgmiiTxChannel (..),
+  rgmiiReceiver,
+  rgmiiSender,
+  unsafeRgmiiRxC,
+  rgmiiTxC,
+) where
+
+import Clash.Prelude
+
+import Protocols
+import Protocols.PacketStream
+
+import Data.Maybe (isJust)
+
+-- | RX channel from the RGMII PHY
+data RgmiiRxChannel dom domDDR = RgmiiRxChannel
```

Comment:
Please note that our style guide says you don't do more than two uppercase characters in a row in the names, so the domain here should be called `domDdr` rather than `domDDR`.

Reply:
`domDdr` looks ugly so I went with `ddrDom` :)

Reply:
I agree it's ugly 😅. But in other code, we _do_ use the `dom` as a prefix, and I think consistency is more important than beauty. So it'd be my preference to use `domDdr` anyway. You could also pick a different suffix, but I don't know a good one. `Clash.Explicit.DDR` uses `slow` and `fast` (without `dom`), but I think that would look ugly here.